                                     Tecnologico de Costa Rica
                                  
                                       Claudia Alcazar Urena

### EDX 2013 Academic  Year Courses

This release is comprised of de-identified data from the first year (Academic Year 2013: Fall 2012, Spring 2013, and Summer 2013) of MITx and HarvardX courses on the edX platform along with related documentation. These data are aggregate records, and each record represents one individual's activity in one edX course. For more information about the existing analyses of these data and the first year of HarvardX and MITx courses, please see the HarvardX and MITx working paper "HarvardX and MITx: The first year of open online courses" by Andrew Ho, Justin Reich, Sergiy Nesterko, Daniel Seaton, Tommy Mullaney, Jim Waldo, and Isaac Chuang (http://papers.ssrn.com/sol3/papers.cfm?abstract_id=2381263). The first release of this dataset is the HarvardX-MITx Person-Course Academic Year 2013 De-Identified dataset, version 2.0, created on May 14, 2014. File name: HMXPC13_DI_v1_5-14-14.csv The md5sum for this release (HMXPC13_DI_v2_5-14-14.csv) is: 2b09c674af772d45dae429045cf7acfc (2014-05-24)


##### Description

    Course_Id: 
    
Administrative, string, identifies institution (HarvardX or MITx), course name, and semester, e.g. “HarvardX/CB22x/2013_Spring”.

    User_Id_DI: 

Administrative, string, first portion identifies dataset (MHxPC13 corresponds to MITx HarvardX Person-­‐Course AY13), second portion is a random ID number. Example ID: “MHxPC130442623”.

    Registered: 

Administrative, 0/1; registered for course, =1 for all records in person course.

    Viewed: 

Administrative, 0/1; anyone who accessed the ‘Courseware’ tab (the home of the videos, problem sets, and exams) within the edX platform for the course. Note that there exist course materials outside of the ‘Courseware’ tab, such as the Syllabus or the Discussion forums.

    Explored: 
    
Administrative, 0/1; anyone who accessed at least half of the chapters in the courseware (chapters are the highest level on the “courseware” menu housing course content).

    Certified: 

Administrative, 0/1; anyone who earned a certificate. Certificates are based on course grades, and depending on the course, the cutoff for a certificate varies from 50% - 80%.

    Final_CC_Cname_DI: 

Mix of administrative (computed from IP address) and user-­‐ provided (filled in from student address if available when IP was indeterminate); during de-­‐identification, some country names were replaced with the corresponding continent/region name. Examples: “Other South Asia” or “Russian Federation”.

    LoE: 
    
User‐provided, highest level of education completed. Possible values: “Less than Secondary,” “Secondary,” “Bachelor’s,” “Master’s,” and “Doctorate.”

    YoB: 
    
User‐provided, year of birth. Example: “1980”.

    Gender: 
    
User‐provided. Possible values: m (male), f (female) and o (other).

    Grade: 

Administrative, final grade in the course, ranges from 0 to 1. Example: “0.87”.

    Start_Time_DI: 

Administrative, date of course registration. Example: “12/19/12”.

    Last_Event_DI: 

Administrative, date of last interaction with course, blank if no interactions beyond registration. Example “11/17/13”.

    NEvents: 

Administrative, number of interactions with the course, recorded in the tracking logs; blank if no interactions beyond registration. Example: “502”.

    NDays_Act: 

Administrative, number of unique days student interacted with course. Example: “16”.

    NPlay_Video: 

Administrative, number of play video events within the course. Example: “52”.

    NChapters: 

Administrative, number of chapters (within the Courseware) with which the student interacted. Example: “12”.

    NForum_Posts: 
    
Administrative, number of posts to the Discussion Forum. Example: “8”.

    Roles: 

Administrative, identifies staff and instructors, but blank as staff and instructors were removed from this release.

    Inconsistent_Flag: 

Administrative, identifies records that are internally inconsistent. Due to a variety of data issues, including missing tracking logs and one course


Referencia: https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/26147

In [39]:
#Importar Librerias

import pyspark
import findspark

In [40]:
findspark.init('C:\Program Files (x86)\Spark')

In [41]:
#Spark Session

spark=SparkSession.builder.appName('tarea_1').getOrCreate()

In [47]:
#Cargar Data
df = spark.read.format('csv').options(header='true', inferSchema='true').load('EdX_2013 Academic Year Courses.csv')

In [163]:
print("El total es:")
df.count()

El total es:


641138

In [164]:
print("El total de columnas es:")
len(df.columns)

El total de columnas es:


23

In [165]:
print("El esquema es:")

df.printSchema()

El esquema es:
root
 |-- course_id: string (nullable = true)
 |-- Course Short Title: string (nullable = true)
 |-- Course Long Title: string (nullable = true)
 |-- userid_DI: string (nullable = true)
 |-- registered: integer (nullable = true)
 |-- viewed: integer (nullable = true)
 |-- explored: integer (nullable = true)
 |-- certified: integer (nullable = true)
 |-- Country: string (nullable = true)
 |-- LoE_DI: string (nullable = true)
 |-- YoB: string (nullable = true)
 |-- Age: string (nullable = true)
 |-- gender: string (nullable = true)
 |-- grade: string (nullable = true)
 |-- start_time_DI: string (nullable = true)
 |-- last_event_DI: string (nullable = true)
 |-- nevents: integer (nullable = true)
 |-- ndays_act: integer (nullable = true)
 |-- nplay_video: integer (nullable = true)
 |-- nchapters: integer (nullable = true)
 |-- nforum_posts: integer (nullable = true)
 |-- roles: string (nullable = true)
 |-- incomplete_flag: integer (nullable = true)



### Parte II - Ejercicios Vistos en Clase

### Sample of the DataFrame

In [184]:
print("Seleccionar solo las columnas a utilizar")

df_sample = df.select("course_id", "userid_DI","registered","viewed","explored", "certified","Country","Age", "gender","grade")

Seleccionar solo las columnas a utilizar


### Show Results 

In [167]:
print("Mostrar primeros 10 Registros")

df_sample.show(10)

Mostrar primeros 10 Registros
+--------------------+--------------+----------+------+--------+---------+-------------+-------+------+-----+
|           course_id|     userid_DI|registered|viewed|explored|certified|      Country|    Age|gender|grade|
+--------------------+--------------+----------+------+--------+---------+-------------+-------+------+-----+
|HarvardX/CB22x/20...|MHxPC130442623|         1|     0|       0|        0|United States|#VALUE!|    NA|    0|
|HarvardX/CB22x/20...|MHxPC130275857|         1|     0|       0|        0|United States|#VALUE!|    NA|    0|
|HarvardX/CB22x/20...|MHxPC130539455|         1|     1|       0|        0|       France|#VALUE!|    NA|    0|
|HarvardX/CB22x/20...|MHxPC130088379|         1|     1|       0|        0|United States|#VALUE!|    NA|    0|
|HarvardX/CB22x/20...|MHxPC130024894|         1|     1|       0|        0|United States|#VALUE!|    NA| 0.07|
|HarvardX/CB22x/20...|MHxPC130340959|         1|     1|       0|        0|United States|#V

### Analisis de Valor del Archivo 

#### 1. ¿Cuantos usuarios estan certificados por pais?

In [124]:
df_sample.where(df_sample['certified'] == 1).groupBy("Country").count().sort("count", ascending=True).show(30)

+--------------------+-----+
|             Country|count|
+--------------------+-----+
|       Other Oceania|    5|
|             Morocco|   28|
|          Bangladesh|   34|
|               Japan|   40|
|       Unknown/Other|   61|
|               China|   62|
|         Philippines|   81|
|           Indonesia|   96|
|            Portugal|  112|
|               Egypt|  120|
|     Other East Asia|  127|
|             Nigeria|  143|
|            Pakistan|  157|
|              Mexico|  168|
|Other North & Cen...|  169|
|           Australia|  196|
|            Colombia|  202|
|             Ukraine|  203|
|              France|  204|
| Other South America|  292|
|              Greece|  317|
|Other Middle East...|  320|
|              Canada|  333|
|    Other South Asia|  408|
|              Poland|  413|
|              Brazil|  453|
|             Germany|  462|
|        Other Africa|  599|
|  Russian Federation|  635|
|               Spain|  837|
+--------------------+-----+
only showing t

In [141]:
df_sample.where(df_sample['certified'] == 1).groupBy("Country").count().agg(sum("count")).collect()

[Row(sum(count)=17687)]

El total de usuarios certificados por el EDX durante el 2013 fue de 17687, tanto hombres como mujeres.

#### 2. ¿Cuantos usuarios que estan certificados por pais son mujeres?

In [142]:
df_sample.where((df_sample['certified'] == 1) & (df_sample['gender'] == "f")).groupBy("Country").count().sort("count", ascending=True).show(100)

+--------------------+-----+
|             Country|count|
+--------------------+-----+
|       Other Oceania|    1|
|               Japan|    2|
|             Morocco|    3|
|          Bangladesh|    5|
|       Unknown/Other|   17|
|         Philippines|   21|
|               China|   22|
|            Portugal|   24|
|            Pakistan|   25|
|             Ukraine|   27|
|              Mexico|   30|
|     Other East Asia|   32|
|               Egypt|   32|
|             Nigeria|   34|
|              Poland|   37|
|Other North & Cen...|   41|
|           Indonesia|   45|
|              France|   46|
|  Russian Federation|   51|
|            Colombia|   52|
|           Australia|   53|
| Other South America|   64|
|              Brazil|   77|
|              Canada|   85|
|Other Middle East...|   90|
|    Other South Asia|   97|
|             Germany|  107|
|              Greece|  115|
|        Other Africa|  158|
|               Spain|  182|
|      United Kingdom|  211|
|        Other

In [140]:
df_sample.where((df_sample['certified'] == 1) & (df_sample['gender'] == "f")).groupBy("Country").count().agg(sum("count")).collect()

[Row(sum(count)=4232)]

Del total de 17687 solo un total de 4232 mujeres han sido activas en la plataforma de EDX y han recibido al menos una certificacion.

#### 3. ¿Cuantos de los usuarios que estan certificados por pais son mujeres mayores a 35 anios?

In [139]:
df_sample.where((df_sample['certified'] == 1) & (df_sample['gender'] == "f") & (df_sample['age'] > 35)).groupBy("Country").count().sort("count", ascending=True).show(100)

+--------------------+-----+
|             Country|count|
+--------------------+-----+
|     Other East Asia|    1|
|               Japan|    1|
|Other North & Cen...|    1|
|               China|    1|
|             Nigeria|    2|
|               Egypt|    2|
|           Australia|    3|
|  Russian Federation|    4|
|              Brazil|    4|
| Other South America|    4|
|Other Middle East...|    5|
|            Colombia|    5|
|    Other South Asia|    8|
|             Germany|    8|
|              Canada|   11|
|      United Kingdom|   16|
|        Other Africa|   19|
|              Greece|   19|
|               Spain|   21|
|               India|   53|
|        Other Europe|   76|
|       United States|  418|
+--------------------+-----+



In [138]:
df_sample.where((df_sample['certified'] == 1) & (df_sample['gender'] == "f") & (df_sample['age'] > 35)).groupBy("Country").count().agg(sum("count")).collect()

[Row(sum(count)=682)]

Es posible determinar que de 4232 solo 682 de las mujeres presentes en este tipo de curso son mayores a 35 anios.

#### 4. ¿Cuales son los cursos donde hubo mayor presencia femenina que obtuvieron una certificacion por parte del EDX?

In [132]:
df_sample.where((df_sample['certified'] == 1) & (df_sample['gender'] == "f")).groupBy("course_id").count().sort("count", ascending=False).show(50)

+--------------------+-----+
|           course_id|count|
+--------------------+-----+
|MITx/14.73x/2013_...| 1107|
|HarvardX/PH207x/2...|  814|
|HarvardX/ER22x/20...|  711|
|HarvardX/PH278x/2...|  298|
|MITx/6.00x/2012_Fall|  242|
|MITx/7.00x/2013_S...|  226|
| HarvardX/CS50x/2012|  162|
|HarvardX/CB22x/20...|  147|
|MITx/6.002x/2012_...|  134|
|MITx/6.00x/2013_S...|  123|
|MITx/3.091x/2012_...|  121|
|MITx/6.002x/2013_...|   52|
|MITx/8.02x/2013_S...|   45|
|MITx/8.MReV/2013_...|   21|
|MITx/3.091x/2013_...|   19|
|MITx/2.01x/2013_S...|   10|
+--------------------+-----+



Un numero total de 1107 de las mujeres que tienen una certificacion por parte del EDX cursaron "Los Desafios de la Pobreza Mundial"(MITx/14.73x/).

#### 5. ¿Cuales son las notas promedio por curso donde hubo mayor presencia femenina que obtuvieron una certificacion por parte del EDX?

In [137]:
df_sample.where((df_sample['certified'] == 1) & (df_sample['gender'] == "f")).groupBy("course_id").agg(avg('grade')*100).show(50)

+--------------------+------------------+
|           course_id|(avg(grade) * 100)|
+--------------------+------------------+
| HarvardX/CS50x/2012|             100.0|
|MITx/6.002x/2012_...| 88.55970149253727|
|MITx/6.002x/2013_...| 78.94230769230771|
|MITx/2.01x/2013_S...| 71.50000000000001|
|MITx/3.091x/2013_...| 81.10526315789474|
|MITx/8.02x/2013_S...|  74.2222222222222|
|HarvardX/CB22x/20...| 76.09523809523814|
|HarvardX/ER22x/20...| 81.74542897327723|
|MITx/8.MReV/2013_...|  73.7142857142857|
|MITx/3.091x/2012_...| 85.71074380165287|
|MITx/6.00x/2013_S...| 77.48780487804876|
|MITx/7.00x/2013_S...| 73.59292035398225|
|MITx/6.00x/2012_Fall| 80.43801652892562|
|HarvardX/PH278x/2...| 74.81879194630872|
|MITx/14.73x/2013_...| 78.97199638663048|
|HarvardX/PH207x/2...| 92.82678132678147|
+--------------------+------------------+



Un numero total de 1107 de las mujeres que tienen una certificacion por parte del EDX cursaron "Los Desafios de la Pobreza Mundial"(MITx/14.73x/) con un promedio de 78.97. Mientras tanto llamado "Computer Science" (HarvardX/CS50x) el cual fue cursado por 162 mujeres tuvo un promedio 100.0

### Nuevos Comandos

#### Repartition

Este comando nos permite crear una division logica de un archivo en bloques mas pequenos para que puedan ser procesados simultaneamente, este metodo ayuda a mejorar el rendimiento de las operaciones de spark.

Hay diferentes tipos de repartition: 

    repartition(Column... partitionExprs)
        retorna un data frame basado en una columna especifica y sus valores

    repartition(int numPartitions)
        retorna un data frame basado en un numero especifico de particiones

In [147]:
#Crear una division logica de male, female, null, etc.
dl_rep1 = df.repartition("gender")

#Crear una division logica de 50 bloques
dl_rep2 = df.repartition(50)

#### Write 

Este comando nos permite volver a escribir los resultados parciales o finales de las transfomaciones a un archivo que puede ser csv, parquet, etc.

In [155]:
df.write.csv("/Users/calcazar/Desktop/Test/Csv")

#### Persist and Unpersist 

Este comando nos permite guardar el data frame en la memoria default designada para su uso posterior, y el unpersist remueve y libera los bloques previamente guardados dando especio a nuevos bloques.

In [160]:
df.persist()

DataFrame[course_id: string, Course Short Title: string, Course Long Title: string, userid_DI: string, registered: int, viewed: int, explored: int, certified: int, Country: string, LoE_DI: string, YoB: string, Age: string, gender: string, grade: string, start_time_DI: string, last_event_DI: string, nevents: int, ndays_act: int, nplay_video: int, nchapters: int, nforum_posts: int, roles: string, incomplete_flag: int]

In [161]:
df.unpersist()

DataFrame[course_id: string, Course Short Title: string, Course Long Title: string, userid_DI: string, registered: int, viewed: int, explored: int, certified: int, Country: string, LoE_DI: string, YoB: string, Age: string, gender: string, grade: string, start_time_DI: string, last_event_DI: string, nevents: int, ndays_act: int, nplay_video: int, nchapters: int, nforum_posts: int, roles: string, incomplete_flag: int]